src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ConditionalOperator.cs

```diff
@@ -2,16 +2,17 @@
 
 using System.Diagnostics;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
-using Microsoft.CodeAnalysis.CSharp.Syntax;
-using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
     internal sealed partial class LocalRewriter
     {
         /// <summary>
-        /// If the condition has a constant value, then just use the selected branch.
-        /// e.g. "true ? x : y" becomes "x".
+        /// If the conditional operator matches to one of the following patterns replaces it with more optimal expression.
+        /// "true ? x : y" becomes "x"
+        /// "false" ? x : y" becomes "y"
+        /// "condition ? true : false" becomes "condition"
+        /// "condition ? false: true" becomes "!condition"
```

Comment:
can these optimizations not be done in the JIT?  then all languages will benefit, not just C#.

Reply:
Checked out sharplab, seems not

Reply:
My question is: shouldn't these improvements be made to the *jit* instead instead of at a single compiler level?

Reply:
The JIT optimizes the first two already. It doesn't optimize the latter two, but I also wouldn't expect those coding patterns to be very common (vs just returning `condition` and `!condition`).

<https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgWQE8BhAGz1wYG8aDYU0bFySBgEsAdhgYBBABSz5CNNLkM2ASiEjB1EcaYB2BhigBXGAwD8DBAxDaA3PuEBfDwx/0mEpryAEIqWupB2npGBj7GxOYAZthcuLYOTi5s7jFePn5igcAQEFwMHErFpQyQMgAmUhhSEDLRxoYmIgk1LQ1NLfYW1rYuyakwOcbeub4z/uKSVWUAIpUlZbV9za0+HZ1mPfWN24Njac5DNpMi055AA>

However, at the same time the JIT is impacted by the overall IL size of emitted methods (and this isn't an easy problem to fix due to a lot of complicated reasons, namely the cost of resolving certain tokens during importation), so the C# compiler getting these optimizations and producing smaller IL would still be beneficial.

A post-compilation general-purpose IL optimization tool would also allow some of these optimizations and would be cross-platform, but it has significantly higher cost to implementing and maintaining. It might be a good idea for a dotnet-foundation project/experiment that could eventually be productized.

Reply:
Actually, looking at the IL, it looks like the C# compiler is already doing the first two itself as well (the JIT still handles it as well due to its constant folding support): <https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BZATwGFcBDAZ27oG8qdYQ3qFSSOtgB2GOgEEAFDLkI0U2XRYBKISMGURRhgHY6GKAFcYdAPx0EdEFoDce4QF93db7QYSNOQAhZU01QK1dQ31vI0IzADNOXG4be0dnFjdoz29fMQDgCAhcOjZFIpK6SGkAE2wMbAhpKKMDYxF46ub6xua7cysbZySUmGyjLxyfab9xSUrSgBEK4tKa3qaW73aO0266hq2B0dSnQesJkSmPIA==>